### DISCUSSION TITLE
Build error about material-insiders version 9.5.3+insiders.4.48.0

### DISCUSSION THREAD

**Discussion by baytogi** [squidfunk/mkdocs-material#6569] [Q&A] [answered]
Hello all,
i am struggling about some issue which occurs only for "insiders" package. Here is the situation,
1- i am using the latest insiders version and pyhton 3.12.1
2- i can't build my website i am getting error like this. See the outputs.
![1](https://github.com/squidfunk/mkdocs-material/assets/7778140/6970e078-69fb-496c-8308-bd6c8eec9657)
3- If i install the non-insiders package than i am able to build successfully. See the outputs.
![2](https://github.com/squidfunk/mkdocs-material/assets/7778140/8e73deb0-3035-4119-9821-7230d7ab1990)

Why i am getting this error when i switch to the insiders version?

**Comment by squidfunk** [comment]
Thanks for reporting. Could you please create a [minimal reproduction](https://squidfunk.github.io/mkdocs-material/guides/creating-a-reproduction/)? If you could provide us with a runnable artifact, we can investigate and help you much more easily.

**Reply by alexvoss** [reply]
It might also help to try running the minimal reproduction outside Google Drive, in case that is what is getting in the way? We might not be able to reproduce this if it is something as specific as running on a Google mapped drive on Windows. (I am not sure if it is a good idea to run inside a sync-ed folder like Google Drive, OneDrive or Dropbox as this must trigger *a lot* of sync activity?) 

Looking at the error messages, it seems the social plugin is involved, which would explain the difference between the Insiders and public versions. That would make it easier to narrow down where the cause lies.

**Reply by squidfunk** [reply]
It appears to be an issue with `os.path.relpath` on Windows, but we need a reproduction to investigate.

**Reply by kamilkrzyskow** [reply]
The error seems quite self-explanatory:
ValueError: Path on C while start on D drive.

And it would be rather hard to provide a minimal reproduction, as it concerns how the Python environment is setup in relation to the working directory.

https://github.com/squidfunk/mkdocs-material-insiders/blob/548d39f3ba9b5e951a480981ade92f2348d159b5/src/plugins/social/plugin.py#L712

The fallback path is located in the site-packages of his C/AppData local user files, and therefore it's impossible to create a relative path from the working D/GoogleDrive directory to that other one.

The solution would be to use the `dirpath` method or just `path` instead of `relpath` in that instance on line 715 where the error originates from.

**Reply by squidfunk** [reply]
I'm confused. I'm not a Windows user, but I read that this is a problem between different mounts. `os.path.dirpath` doesn't exist (at least my searches don't return anything), so I'm not sure what you mean.

**Reply by kamilkrzyskow** [reply]
Sorry, I tend to use the pathlib nowadays and I would bet my money that there is an easy `pathlib.Path(__file__).parent` `os.path` equivalent.
However, after checking online it seems that this is as simple as it will get:
`os.path.dirname(os.path.abspath(__file__))`, and should yield a base absolute path to the directory.

As for the different mounts, each partition in Windows acts as different mounts, so yes that's the issue.

EDIT:
I did some research and it turned out that only since Python 3.9, the `__file__` path is absolute, and therefore even `pathlib.Path.parent` wouldn't work as I expected in Python 3.8. Here is an example from my Windows machine and Git Bash:

<details>
<summary>Running from same partition and working directory</summary>

```py
/c/MyFiles $ python312 s.py
__file__ = 'C:\\MyFiles\\s.py'
os.path.dirname(__file__) = 'C:\\MyFiles'
os.path.dirname(os.path.abspath(__file__)) = 'C:\\MyFiles'
pathlib.Path(__file__).parent = WindowsPath('C:/MyFiles')
pathlib.Path(__file__).absolute().parent = WindowsPath('C:/MyFiles')
/c/MyFiles $ python38 s.py
__file__ = 's.py'
os.path.dirname(__file__) = ''
os.path.dirname(os.path.abspath(__file__)) = 'C:\\MyFiles'
pathlib.Path(__file__).parent = WindowsPath('.')
pathlib.Path(__file__).absolute().parent = WindowsPath('C:/MyFiles')
```

</details>

<details>
<summary>Running from different partition/mount</summary>

```py
/d $ python312 /c/MyFiles/s.py
__file__ = 'C:\\MyFiles\\s.py'
os.path.dirname(__file__) = 'C:\\MyFiles'
os.path.dirname(os.path.abspath(__file__)) = 'C:\\MyFiles'
pathlib.Path(__file__).parent = WindowsPath('C:/MyFiles')
pathlib.Path(__file__).absolute().parent = WindowsPath('C:/MyFiles')
/d $ python38 /c/MyFiles/s.py
__file__ = 'C:/MyFiles/s.py'
os.path.dirname(__file__) = 'C:/MyFiles'
os.path.dirname(os.path.abspath(__file__)) = 'C:\\MyFiles'
pathlib.Path(__file__).parent = WindowsPath('C:/MyFiles')
pathlib.Path(__file__).absolute().parent = WindowsPath('C:/MyFiles')
```

</details>

---

Based on this my recommendation would be, for any base path created via `os.path.dirname(__file__)` to be changed to `os.path.dirname(os.path.abspath(__file__))`, as it will assure same behaviour on Python 3.8 and above.
So modify line 712 https://github.com/squidfunk/mkdocs-material-insiders/blob/548d39f3ba9b5e951a480981ade92f2348d159b5/src/plugins/social/plugin.py#L712
and use simply `path` on line 715

**Reply by squidfunk** [reply]
Thanks for investigating! I'm confused about your edit – does the recommendation following the edit still work? I have no means of testing this (not a Windows user), so if you can actually test this and spare the time to PR' the change, that would be amazing!

**Reply by kamilkrzyskow** [reply]
The recommendation was based on the edit, and is a part of it 😄. It's possible to check what was changed in each iteration of the message. Anyways, I think that is a change small enough that it doesn't require more testing, but if you prefer to wait for the PR from a Windows user, I think I should be able to test it a bit later today ✌️

**Reply by baytogi** [reply]
Do i still need to test with minimal reproduction or did you understand the issue about insiders package build error? It seems you have an idea about the situation.

**Reply by kamilkrzyskow** [reply]
No need for the reproduction. A simple 
```yml
site_name: My Docs

theme:
  name: material
  
plugins:
  - search
  - social
```
config, together with launching it from another partition, is enough to trigger the crash.

My recommendation above does indeed fix the crash. I'm currently fighting my own PC, because for some reason `npm start` started to create symlinks in the `material/` path and they point to `src/`. So when I try to copy the plugin files from src to material (which are 1:1) it doesn't let me as it's the same file... I also don't want to remove the symlink, as it wasn't like that before. In the past it wasn't so difficult to commit 2 files 😞. However, I might suck it up and adhere to the rules of modifying only `src/` 🤷

**Reply by squidfunk** [reply]
[We've switched to symlinks for `.py` files](https://github.com/squidfunk/mkdocs-material-insiders/blob/01822b145378ead94daa70f065ca4a21286548a3/tools/build/index.ts#L161-L184), because we copy the files to the `material` folder as well, and when working on the project and clicking on a class or function name, you get redirected to the built files. This led to the problem where changes were made to built files and then overwritten, which was annoying. Therefore, `.py` files are now symlinked.

This only happens when you do `npm start`. On `npm run build:all`, files are copied.

You can modify only `src` for the PR, and we can check over at Insiders how we can improve the developer experience on Windows. I'm not sure why this is a problem, symlinks should be possible, shouldn't they?

**Reply by kamilkrzyskow** [reply]
I've created the PR https://github.com/squidfunk/mkdocs-material-insiders/pull/78
As for my issue with the symlinks, I didn't want to add something that I didn't change to the commit.
Running `git status` reveals:
```bash
Changes not staged for commit:
  (use "git add <file>..." to update what will be committed)
  (use "git restore <file>..." to discard changes in working directory)
...
typechange: material/plugins/social/plugin.py
...
```
And when I try to `cp src/plugins/social/plugin.py material/plugins/social/plugin.py` it doesn't let me as it's the same file.
I felt that the change is intended, but I didn't understand why it was made, so thanks for the explanation.
Basically my issue was the inability of just `git add material/plugins/social/plugin.py`, because it would change it to a symlink in the commit, which I assume is invalid, as `material` is used for `pip` installation and there can't be symlinks inside 🤔 
I made the changes in the `src/` path in the PR, might run `npm run build:all` in the future or delete the symlink and copy over the files from `src/` (which might also not work, as now py files aren't minified/stripped of comments and they might as well could be).

**Reply by kamilkrzyskow** [reply]
I forgot to mention that running `npm start` now with plain user privileges results in an error:
```js
$ npm start

> mkdocs-material@9.5.3+insiders-4.49.0 start
> ts-node -T tools/build --verbose --all --dirty --watch

14:36:52 + material/overrides\assets\stylesheets\custom.css
14:36:52 + material/overrides\assets\stylesheets\home.css
14:36:52 + material/templates\assets\stylesheets\palette.css
Error: EPERM: operation not permitted, symlink '..\..\..\src\plugins\blog\author.py' -> 'C:\MyFiles\_git\mkdocs-material-insiders\material\plugins\blog\author.py'
    at async Object.symlink (node:internal/fs/promises:993:10) {
  errno: -4048,
  code: 'EPERM',
  syscall: 'symlink',
  path: '..\\..\\..\\src\\plugins\\blog\\author.py',
  dest: 'C:\\MyFiles\\_git\\mkdocs-material-insiders\\material\\plugins\\blog\\author.py'
}
```
By default Windows 10 requires administrative privileges for symlink creation, as it's seen as a security vulnerability:
https://www.stigviewer.com/stig/windows_10/2020-06-15/finding/V-63865

One option and obvious solution is running the terminal as administrator and then running `npm start`, but I guess people might be opposed to the idea 🤔 
Another is to enable `developer mode` https://stackoverflow.com/a/58048888 Run `start ms-settings:developers` in `cmd`
Another is to enable symlinks for normal users in Group Policies `Computer Configuration >> Windows Settings >> Security Settings >> Local Policies >> User Rights Assignment >> "Create symbolic links"`

Further read: https://github.com/npm/npm/issues/10926#issuecomment-199546531

**Reply by squidfunk** [reply]
We can limit symbolic links to non-Windows systems for now, adding `process.platform !== "win32"`. That should alleviate the problems until we find a better solution for the build system. I'm going to rework the build system at some point, but it's currently not a priority. However, I agree that we should ensure building works on all systems.

**Reply by baytogi** [reply]
Hi all, i am still waiting for a new release or suggestion about handling my issue. What is your suggestion for my insider building? i need to update my wiki page my i cant do it right now. Is there any workover for me?

**Reply by kamilkrzyskow** [reply]
Hi @baytogi,
a quick workaround for you would be to not use `D:\0-GoogleDrive` the issue comes from 2 different mount paths for the plugin internals and your files, due to a bug in the code. So you can just move the directory to the `C:` drive and everything should work as expected.
The PR should be merged into Insiders today or tomorrow after I add a small thing to it.
EDIT: `C:\` has some issues with permissions on most Windows systems, so perhaps better use your `C:\Users\baytogi\Documnets` directory. You get the idea :v:
EDIT2: You can also create a Virtual Environment https://docs.python.org/3/tutorial/venv.html in the `D:` partition and activate it when using `D:\0-GoogleDrive`

**Reply by squidfunk** [reply, marked as answer]
I just merged the PR, so you can try the latest `master`.

**Reply by baytogi** [reply]
Thanks for your replies. Latest merged PR solved the issue.
